src/llvm/abi/wasm32.cr

```diff
@@ -0,0 +1,64 @@
+require "../abi"
+
+class LLVM::ABI::WebAssembly32 < LLVM::ABI
```

Comment:
This looks like perhaps it's the ARM ABI? I think there's a few things not quite right, based on the [rust one](https://github.com/rust-lang/rust/blob/master/compiler/rustc_target/src/abi/call/wasm.rs). Since WASM doesn't have registers, we don't need to do any of the stuff around that or worrying too much about sizes. We can just extend everything to 4 bytes for numbers and pointer types, and pass everything else by value (not sure if we need to do anything like `unwrap_trivial_aggregate` or  if that's already handled elsewhere)

This was the approximation of the Rust implementation devised earlier from the one in the original [asm.js port](https://github.com/crystal-lang/crystal/pull/3634/files#diff-38fdb515276d64790d609086a28fc99d37551a13ee48861769cf3d06fa9de43e): https://raw.githubusercontent.com/maxfierke/crystal/mf-spike_wasm_take_two/src/llvm/abi/wasm32.cr, which lines up a bit better with what WASM expects

Reply:
I went ahead to use your ABI and it makes a lot more sense, thank!! I was using the ARM one because it worked and I didn't know enough LLVM to do my own.